### PR TITLE
initial commit of implementation NGenius payment intergation

### PR DESCRIPTION
**Pull request by DFSko**

### CODE REVIEW

**Comment by mingazhev** [comment]
will postpone this for now

**Review by mingazhev — COMMENTED** [review]
Adding a comment to remove from PR's to review

File: Api/Controllers/AgentControllers/PaymentsController.cs [hunk 1]
```diff
@@ -41,6 +44,22 @@ public class PaymentsController : BaseController
         public IActionResult GetCurrencies() => Ok(_paymentSettingsService.GetCurrencies());
 
 
+        /// <summary>
+        ///     Returns available payment systems
+        /// </summary>>
+        [HttpGet("payment-systems")]
+        [ProducesResponseType(typeof(IReadOnlyCollection<PaymentSystems>), StatusCodes.Status200OK)]
+        public IActionResult GetPaymentSystems() => Ok(_paymentSettingsService.GetPaymentSystems());
```
Comment (mingazhev) [inline Api/Controllers/AgentControllers/PaymentsController.cs:52]: Arrow to the next line

File: Api/Controllers/AgentControllers/PaymentsController.cs [hunk 2]
```diff
@@ -41,6 +44,22 @@ public class PaymentsController : BaseController
...
+
+
+        /// <summary>
+        ///     Returns enabled payment system
+        /// </summary>>
+        [HttpGet("payment-system")]
+        [ProducesResponseType(typeof(PaymentSystems), StatusCodes.Status200OK)]
+        public IActionResult GetEnabledPaymentSystem() => Ok(_paymentSettingsService.GetEnabledPaymentSystem());
```
Comment (mingazhev) [inline Api/Controllers/AgentControllers/PaymentsController.cs:60]: `GetCurrentPaymentSystem`

File: Api/appsettings.json [hunk 3]
```diff
@@ -124,5 +125,8 @@
   },
   "MongoDB": {
     "Options": "edo/mongodb"
+  },
+  "PaymentSystem": {
+    "EnabledPaymentSystem": "NGenius"
```
Comment (mingazhev) [inline Api/appsettings.json:130]: I think there should be Payfort as default

File: Api/Services/Payments/NGenius/AuthResponse.cs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+namespace HappyTravel.Edo.Api.Services.Payments.NGenius
+{
+    public readonly struct AuthResponse
```
[the remark this answers is not in the document]
Comment (mingazhev) [reply Api/Services/Payments/NGenius/AuthResponse.cs:5]: This looks like to be in Models

File: Api/Services/Payments/NGenius/Models/BillingAddress.cs [hunk 5]
```diff
@@ -0,0 +1,8 @@
+namespace HappyTravel.Edo.Api.Services.Payments.NGenius.Models
+{
+    public readonly struct BillingAddress
```
Comment (kirillta) [inline Api/Services/Payments/NGenius/Models/BillingAddress.cs:3]: Why that's an address if it contains only names?

File: Api/Services/Payments/NGenius/Models/NGeniusPaymentResponse.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
+using HappyTravel.Edo.Common.Enums;
+
+namespace HappyTravel.Edo.Api.Services.Payments.NGenius.Models
```
Comment (kirillta) [inline Api/Services/Payments/NGenius/Models/NGeniusPaymentResponse.cs:3]: Models within services?

File: Api/Services/Payments/NGenius/Models/Payment.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+namespace HappyTravel.Edo.Api.Services.Payments.NGenius.Models
+{
+    public readonly struct Payment
+    {
+        public string Pan { get; init; }
+        public string Expiry { get; init; }
+        public string Cvv { get; init; }
+        public string CardholderName { get; init; }
```
Comment (kirillta) [inline Api/Services/Payments/NGenius/Models/Payment.cs:8]: `CardHolderName` I believe

File: Api/Services/Payments/NGenius/Models/ResponsePaymentInformation.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
+namespace HappyTravel.Edo.Api.Services.Payments.NGenius.Models
+{
+    public readonly struct ResponsePaymentInformation
```
Comment (kirillta) [inline Api/Services/Payments/NGenius/Models/ResponsePaymentInformation.cs:3]: This model is quite similar with Payment.cs

File: Api/Services/Payments/NGenius/Models/Secure3dOptions.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+        {
+            AcsUrl = acsUrl;
+            AcsPaReq = acsPaReq;
+            AcsMd = acsMd;
+        }
+
+
+        public string AcsUrl { get; }
+        public string AcsPaReq { get; }
+        public string AcsMd { get; }
```
Comment (kirillta) [inline Api/Services/Payments/NGenius/Models/Secure3dOptions.cs:15]: Do these parameters have meaningful names?

File: Api/Services/Payments/NGenius/NGeniusHttpClient.cs [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+        }
+
+
+        private async Task<string> GetAccessToken()
+        {
+            var key = _cache.BuildKey(nameof(NGeniusHttpClient), "access-token");
+            
+            if (!_cache.TryGetValue<string>(key, out var token))
+            {
+                var request = new HttpRequestMessage(HttpMethod.Post, "identity/auth/access-token");
```
Comment (kirillta) [inline Api/Services/Payments/NGenius/NGeniusHttpClient.cs:37]: `HttpRequestMessage` and `HttpResponseMessage` are disposable

File: Api/Services/Payments/NGenius/NGeniusHttpClient.cs [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+                var data = JsonSerializer.Deserialize<AuthResponse>(await response.Content.ReadAsStringAsync());
+                _cache.Set(key, data.AccessToken, TimeSpan.FromSeconds(data.ExpiresIn).Subtract(TimeSpan.FromMinutes(1)));
+                token = data.AccessToken;
+            }
+
+            return token;
+        }
+
+
+        public async Task<Result<NGeniusPaymentResponse>> Authorize(OrderRequest order)
```
Comment (kirillta) [inline Api/Services/Payments/NGenius/NGeniusHttpClient.cs:52]: Could you make this method simplier?

File: Api/Services/Payments/NGenius/NGeniusPaymentService.cs [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+                var (_, isFailure, error) = await paymentCallbackService.ProcessPaymentChanges(payment);
+
+                return isFailure
+                    ? Result.Failure<NGeniusPaymentResponse>(error)
+                    : Result.Success(paymentResult);
+            }
+        }
+        
+        
+        public async Task<Result<NGeniusPaymentResponse>> Authorize(SavedCreditCardRequest request, string languageCode, string ipAddress, IPaymentCallbackService paymentCallbackService, AgentContext agent)
```
Comment (kirillta) [inline Api/Services/Payments/NGenius/NGeniusPaymentService.cs:81]: These two method do pretty much the same. Could they be generalized?

**Comment by DFSko** [comment]
implemented in another branches

**Pull request closed without merging**